### PR TITLE
Data path traffic separation option in swarm mode

### PR DESCRIPTION
**Pull request by fcrisciani**
**- What I did**
In SWARM mode is now possible to pass a new flag to specify which is the interface to use for data path traffic. This will allow to separate data traffic (container to container pkts) from control traffic (grpc between workers and managers)

**- How I did it**
Introduced a new flag --datapath-addr in the docker swarm init and docker swarm join

**- How to verify it**
Create a swarm cluster on hosts with at least 2 interfaces, one will be dedicated for the management and will be the one that will be used in the --advertise-addr, the other instead will be the one for the --datapath-addr flag.
Second step is to create an overlay network
Third step is to launch containers on different workers using the overlay network.
At this point the traffic between the containers on different workers will happen on the interface identified by the --datapath-addr

**- Description for the changelog**

Data path traffic separation option in swarm mode

**- A picture of a cute animal (not mandatory but encouraged)**
hqdefault

### CODE REVIEW

File: daemon/cluster/cluster.go [hunk 1]
```diff
@@ -270,6 +270,16 @@ func (c *Cluster) GetAdvertiseAddress() string {
 	return c.currentNodeState().actualLocalAddr
 }
 
+// GetDataPathAddress returns the address to be used for the data path traffic.
+func (c *Cluster) GetDataPathAddress() string {
```
Comment (aboch) [inline daemon/cluster/cluster.go]: I think you no longer need this function

File: daemon/cluster/listen_addr.go [hunk 2]
```diff
@@ -103,6 +88,23 @@ func (c *Cluster) resolveAdvertiseAddr(advertiseAddr, listenAddrPort string) (st
 	return systemAddr.String(), listenAddrPort, nil
 }
 
+func resolveDataPathAddr(dataPathAddr string) (string, error) {
+	// Approach:
+	// - If a data path flag is specified try to resolve the IP address.
+	if dataPathAddr != "" {
```
Comment (aboch) [inline daemon/cluster/listen_addr.go]: I would handle the empty case right away and avoid the extra indentation here with 
```
if dataPathAddr == "" {
    return "", nil
}
```

File: daemon/cluster/cluster.go (hunk 1 above)
Comment (aboch) [inline daemon/cluster/cluster.go]: Take it back, this is actually needed when libnetwork retrieves the addresss from the clusterProvider

File: contrib/completion/zsh/_docker [hunk 3]
```diff
@@ -2267,6 +2267,7 @@ __docker_swarm_subcommand() {
             _arguments $(__docker_arguments) \
                 $opts_help \
                 "($help)--advertise-addr=[Advertised address]:ip\:port: " \
+                "($help)--datapath-addr=[Data path ip or interface]:ip " \
```
Comment (aaronlehmann) [inline contrib/completion/zsh/_docker]: I'd make IP uppercase (in "Data path ip...") to be consistent with other descriptions.

File: contrib/completion/zsh/_docker [hunk 4]
```diff
@@ -2282,6 +2283,7 @@ __docker_swarm_subcommand() {
             _arguments $(__docker_arguments) -A '-*' \
                 $opts_help \
                 "($help)--advertise-addr=[Advertised address]:ip\:port: " \
+                "($help)--datapath-addr=[Data path ip or interface]:ip " \
```
Comment (aaronlehmann) [inline contrib/completion/zsh/_docker]: I'd make IP uppercase (in "Data path ip...") to be consistent with other descriptions.

File: docs/reference/commandline/swarm_init.md [hunk 5]
```diff
@@ -118,6 +119,12 @@ for example `--advertise-addr eth0:2377`.
 Specifying a port is optional. If the value is a bare IP address or interface
 name, the default port 2377 will be used.
 
+### `--datapath-addr`
+
+This flag specifies the address or the interface that is going to be used for
+data path traffic in global scope drivers. If unspecified, Docker will use the
+same IP address and interface that is used for the advertise address.
```
Comment (aaronlehmann) [inline docs/reference/commandline/swarm_init.md]: Please clarify in the documentation that this is an address that's used by other nodes, rather than one used for local binding.

File: daemon/cluster/listen_addr.go [hunk 6]
```diff
@@ -149,6 +151,32 @@ func resolveInterfaceAddr(specifiedInterface string) (net.IP, error) {
...
+		return interfaceAddr, nil
+	}
+	// String matched interface but there is a potential ambiguity to be resolved
+	if err != errNoSuchInterface {
+		return nil, err
+	}
+
+	// String is not an interface check if it is a valid IP
+	if ip := net.ParseIP(input); ip != nil {
+		if isUnspecifiedValid || !ip.IsUnspecified() {
```
Comment (aaronlehmann) [inline daemon/cluster/listen_addr.go]: These conditions could be folded into the outer `if`.

```go
if ip := net.ParseIP(input); ip != nil && (isUnspecifiedValid || !ip.IsUnspecified()) {
        return ip, nil
}
```

Or perhaps something like this:

```go
ip := net.ParseIP(input)
if ip == nil {
        return nil, errBadNetworkIdentifier
}
if ip.IsUnspecified()  && !isUnspecifiedValid{
        return nil, errBadNetworkIdentifier
}
return return ip, nil
```

File: integration-cli/docker_cli_swarm_test.go [hunk 7]
```diff
@@ -95,6 +95,13 @@ func (s *DockerSwarmSuite) TestSwarmInitUnspecifiedAdvertiseAddr(c *check.C) {
 	c.Assert(out, checker.Contains, "advertise address must be a non-zero IP address")
 }
 
+func (s *DockerSwarmSuite) TestSwarmInitUnspecifiedDataPathAddr(c *check.C) {
+	d := s.AddDaemon(c, false, false)
+	out, err := d.Cmd("swarm", "init", "--datapath-addr", "0.0.0.0")
+	c.Assert(err, checker.NotNil)
+	c.Assert(out, checker.Contains, "data path address must be a non-zero IP")
```
Comment (aaronlehmann) [inline integration-cli/docker_cli_swarm_test.go]: Perhaps we could also check that a `--datapath-addr` argument with a port number is rejected.

File: cli/command/swarm/opts.go [hunk 8]
```diff
@@ -19,6 +19,7 @@ const (
 	flagDispatcherHeartbeat = "dispatcher-heartbeat"
 	flagListenAddr          = "listen-addr"
 	flagAdvertiseAddr       = "advertise-addr"
+	flagDataPathAddr        = "datapath-addr"
```
Comment (aaronlehmann) [inline cli/command/swarm/opts.go]: Wondering if this should be `data-path-addr`. Not sure either way but wanted to bring it up for discussion.

File: docs/reference/commandline/swarm_init.md [hunk 9]
```diff
@@ -22,6 +22,7 @@ Initialize a swarm
 
 Options:
       --advertise-addr string           Advertised address (format: <ip|interface>[:port])
+      --datapath-addr string            Address or interface to use for data path traffic in the cluster (format: <ip|interface>)
```
Comment (aaronlehmann) [inline docs/reference/commandline/swarm_init.md]: The usage info here should match what's printed by `swarm init --help`. I believe the flags are sorted alphabetically in `swarm init --help`.

File: docs/reference/commandline/swarm_join.md [hunk 10]
```diff
@@ -22,6 +22,7 @@ Join a swarm as a node and/or manager
 
 Options:
       --advertise-addr string   Advertised address (format: <ip|interface>[:port])
+      --datapath-addr string    Data path address (format: <ip|interface>)
```
Comment (aaronlehmann) [inline docs/reference/commandline/swarm_join.md]: The usage info here should match what's printed by `swarm join --help`. I believe the flags are sorted alphabetically in `swarm join --help`.

File: vendor/github.com/docker/libnetwork/agent.go [hunk 11]
```diff
@@ -189,14 +199,16 @@ func (c *controller) agentSetup() error {
...
 	bindAddr := clusterProvider.GetLocalAddress()
 	advAddr := clusterProvider.GetAdvertiseAddress()
+	dataAddr := clusterProvider.GetDataPathAddress()
 	remote := clusterProvider.GetRemoteAddress()
 	remoteAddr, _, _ := net.SplitHostPort(remote)
 	listen := clusterProvider.GetListenAddress()
 	listenAddr, _, _ := net.SplitHostPort(listen)
 
-	logrus.Infof("Initializing Libnetwork Agent Listen-Addr=%s Local-addr=%s Adv-addr=%s Remote-addr =%s", listenAddr, bindAddr, advAddr, remoteAddr)
+	logrus.Infof("Initializing Libnetwork Agent Listen-Addr=%s Local-addr=%s Adv-addr=%s Data-addr=%s Remote-addr =%s",
```
Comment (aaronlehmann) [inline vendor/github.com/docker/libnetwork/agent.go]: Maybe this log message should exclude `Data-addr` if it's an empty string, for better readability.

Unrelated to this change, but the space between `Remote-addr` and `=%s` is inconsistent with the other fields.

File: vendor/github.com/docker/libnetwork/endpoint.go [hunk 12]
```diff
@@ -528,7 +528,8 @@ func (ep *endpoint) sbJoin(sb *sandbox, options ...EndpointOption) error {
 	}()
 
 	if sb.needDefaultGW() && sb.getEndpointInGWNetwork() == nil {
-		return sb.setupDefaultGW()
+		err = sb.setupDefaultGW()
+		return err
```
Comment (aaronlehmann) [inline vendor/github.com/docker/libnetwork/endpoint.go]: Why this change?

Comment (aaronlehmann) [inline vendor/github.com/docker/libnetwork/endpoint.go]: Oh, I see the defer now. That's kind of ugly. If a defer is going to inspect the value of `err`, the function should use a "named return":

```go
func (ep *endpoint) sbJoin(sb *sandbox, options ...EndpointOption) (err error) {
```

Then returning an error will automatically assign it to `err`.

Comment (aboch) [inline vendor/github.com/docker/libnetwork/endpoint.go]: This is part of another change brought in by the vendoring. 
It is to make sure the existing defer function which performs the cleanup in case of `err != nil` is being executed.

Comment (aboch) [inline vendor/github.com/docker/libnetwork/endpoint.go]: Yes, I considered the named returned but I decided to not go for that change because I did not want to add refactoring changes to the fix, to speed up review. And to avoid any collateral bugs  for changes not strictly needed, because that function has some delicate logic. After the next stable release is out, I will consider refactoring that.

File: contrib/completion/bash/docker [hunk 13]
```diff
@@ -3303,7 +3303,7 @@ _docker_swarm_init() {
 
 	case "$cur" in
 		-*)
-			COMPREPLY=( $( compgen -W "--advertise-addr --autolock --availability --cert-expiry --dispatcher-heartbeat --external-ca --force-new-cluster --help --listen-addr --max-snapshots --snapshot-interval --task-history-limit" -- "$cur" ) )
+			COMPREPLY=( $( compgen -W "--advertise-addr --datapath-addr --autolock --availability --cert-expiry --dispatcher-heartbeat --external-ca --force-new-cluster --help --listen-addr --max-snapshots --snapshot-interval --task-history-limit" -- "$cur" ) )
```
Comment (aaronlehmann) [inline contrib/completion/bash/docker]: Needs to be updated to `--data-path-addr`

File: contrib/completion/bash/docker [hunk 14]
```diff
@@ -3339,7 +3339,7 @@ _docker_swarm_join() {
 
 	case "$cur" in
 		-*)
-			COMPREPLY=( $( compgen -W "--advertise-addr --availability --help --listen-addr --token" -- "$cur" ) )
+			COMPREPLY=( $( compgen -W "--advertise-addr --datapath-addr --availability --help --listen-addr --token" -- "$cur" ) )
```
Comment (aaronlehmann) [inline contrib/completion/bash/docker]: Needs to be updated to `--data-path-addr`

File: docs/reference/commandline/swarm_init.md [hunk 15]
```diff
@@ -118,6 +119,15 @@ for example `--advertise-addr eth0:2377`.
 Specifying a port is optional. If the value is a bare IP address or interface
 name, the default port 2377 will be used.
 
+### `--data-path-addr`
+
+This flag specifies the address that global scope drivers are going to publish towards
+other workers in order to reach the containers running on this worker node.
```
Comment (aaronlehmann) [inline docs/reference/commandline/swarm_init.md]: I assume this would apply to managers as well (by default, containers can be scheduled to managers as well as workers), so "workers" should be "nodes" and "worker node" should be "node".

Instead of "going to publish", how about "will publish"?

Maybe "global scope drivers" should be "global scope network drivers", for clarity?

File: integration-cli/docker_cli_swarm_test.go [hunk 16]
```diff
@@ -1760,6 +1760,7 @@ func (s *DockerSwarmSuite) TestSwarmJoinWithDrain(c *check.C) {
 	d1 := s.AddDaemon(c, false, false)
 
 	out, err = d1.Cmd("swarm", "join", "--availability=drain", "--token", token, d.ListenAddr)
+	fmt.Printf("** token:%s, listen:%s\n", token, d.ListenAddr)
```
Comment (aaronlehmann) [inline integration-cli/docker_cli_swarm_test.go]: Please remove the `Printf` before merging.

File: daemon/cluster/listen_addr.go [hunk 17]
```diff
@@ -10,8 +10,10 @@ var (
 	errNoSuchInterface         = errors.New("no such interface")
 	errNoIP                    = errors.New("could not find the system's IP address")
 	errMustSpecifyListenAddr   = errors.New("must specify a listening address because the address to advertise is not recognized as a system address, and a system's IP address to use could not be uniquely identified")
+	errBadNetworkIdentifier    = errors.New("must specify a valid IP address or interface name")
 	errBadListenAddr           = errors.New("listen address must be an IP address or network interface (with optional port number)")
 	errBadAdvertiseAddr        = errors.New("advertise address must be a non-zero IP address or network interface (with optional port number)")
+	errBadDataPathAddr         = errors.New("data path address must be a non-zero IP address or network interface")
```
Comment (aaronlehmann) [inline daemon/cluster/listen_addr.go]: can you add `(without a port number)` to this, like `errBadDefaultAdvertiseAddr`?

File: integration-cli/docker_cli_swarm_test.go (hunk 16 above)
Comment (fcrisciani) [inline integration-cli/docker_cli_swarm_test.go]: done

File: daemon/cluster/listen_addr.go (hunk 17 above)
Comment (fcrisciani) [inline daemon/cluster/listen_addr.go]: done

File: docs/reference/commandline/swarm_init.md (hunk 15 above)
Comment (fcrisciani) [inline docs/reference/commandline/swarm_init.md]: done

**Comment by fcrisciani** [comment]
will close this one because does not reflect the commits update
